File: Book/app/src/main/java/com/example/book/MainActivity.java
package com.example.book;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {
 Button b_00,b_01,b_02,
         b_10,b_11,b_12,
         b_20,b_21,b_22, login;


 Board game;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        b_00=findViewById(R.id.b_00);
        b_01=findViewById(R.id.b_01);
        b_02=findViewById(R.id.b_02);
        b_10=findViewById(R.id.b_10);
        b_11=findViewById(R.id.b_11);
        b_12=findViewById(R.id.b_12);
        b_20=findViewById(R.id.b_20);
        b_21=findViewById(R.id.b_21);
        b_22=findViewById(R.id.b_22);
        login=findViewById(R.id.login_id);

        b_00.setOnClickListener(this);
        b_01.setOnClickListener(this);
        b_02.setOnClickListener(this);
        b_10.setOnClickListener(this);
        b_11.setOnClickListener(this);
        b_12.setOnClickListener(this);
        b_20.setOnClickListener(this);
        b_21.setOnClickListener(this);
        b_22.setOnClickListener(this);

        login.setOnClickListener(this);
        game=Board.getBoard();
    }

    public void checkGameStatus(){
        //Intent intent=new Intent(this,GameBackGroundTask.class);
        //startService(intent);
    }
    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.b_00:
                if(b_00.getText().toString().equals("")) {
                    game.setKey(0, 0);
                    checkGameStatus();
                    Intent intent=new Intent(this,GameBackGroundTask.class);
                    startService(intent);
                    if (game.getValue(0, 0))
                        b_00.setText("x");
                    else
                        b_00.setText("o");
                }
                break;
            case R.id.b_01:
                if(b_01.getText().toString().equals("")) {
                    game.setKey(0, 1);
                    checkGameStatus();
                    if (game.getValue(0, 1))
                        b_01.setText("x");
                    else
                        b_01.setText("o");
                }
                break;
            case R.id.b_02:
                if(b_02.getText().toString().equals("")) {
                    game.setKey(0, 2);
                    checkGameStatus();
                    if (game.getValue(0, 2))
                        b_02.setText("x");
                    else
                        b_02.setText("o");
                }
                break;
            case R.id.b_10:
                if(b_10.getText().toString().equals("")) {
                    game.setKey(1, 0);
                    checkGameStatus();
                    if (game.getValue(1, 0))
                        b_10.setText("x");
                    else
                        b_10.setText("o");
                }
                break;
            case R.id.b_11:
                if(b_11.getText().toString().equals("")) {
                    game.setKey(1, 1);
                    checkGameStatus();
                    if (game.getValue(1, 1))
                        b_11.setText("x");
                    else
                        b_11.setText("o");
                }
                break;
            case R.id.b_12:
                if(b_12.getText().toString().equals("")) {
                    game.setKey(1, 2);
                    checkGameStatus();
                    if (game.getValue(1, 2))
                        b_12.setText("x");
                    else
                        b_12.setText("o");
                }
                break;
            case R.id.b_20:
                if(b_20.getText().toString().equals("")) {
                    game.setKey(2, 0);
                    checkGameStatus();
                    if (game.getValue(2, 0))
                        b_20.setText("x");
                    else
                        b_20.setText("o");
                }
                break;
            case R.id.b_21:
                if(b_21.getText().toString().equals("")) {
                    game.setKey(2, 1);
                    checkGameStatus();
                    if (game.getValue(2, 1))
                        b_21.setText("x");
                    else
                        b_21.setText("o");
                }
                break;
            case R.id.b_22:
                if(b_22.getText().toString().equals("")) {
                    game.setKey(2, 2);
                    checkGameStatus();
                    if (game.getValue(2, 2))
                        b_22.setText("x");
                    else
                        b_22.setText("o");
                }
                break;
            case R.id.login_id:


        }
    }
}

File: AndroidApps/app/src/main/java/com/example/firebaselogin/MainActivity.kt
package com.example.firebaselogin

import android.content.ContentValues.TAG
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.FrameLayout
import android.widget.LinearLayout
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.tooling.preview.Preview
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import com.example.firebaselogin.Home.HomeFragment
import com.example.firebaselogin.ui.theme.FirebaseLoginTheme
import com.google.firebase.FirebaseApp
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.ktx.firestore
import com.google.firebase.ktx.Firebase
import com.example.firebaselogin.databinding.ActivityMainBinding

class MainActivity : ComponentActivity() {
    val db = Firebase.firestore
    val auth = FirebaseAuth.getInstance()
    lateinit var userId: String

    private  lateinit var binding: ActivityMainBinding

    private lateinit var loginViewModel: LoginViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        val llayout: LinearLayout = findViewById(R.id.register_details)


        llayout.visibility = View.INVISIBLE


        val factory = LoginViewModelFactory(auth, db)
        loginViewModel = ViewModelProvider(this, factory).get(LoginViewModel::class.java)

        loginViewModel.loginResult.observe(this, Observer{ success ->
            if(success){
                Toast.makeText(this, "Login Successful", Toast.LENGTH_SHORT).show()
            } else{
                Toast.makeText(this, "Login Failed", Toast.LENGTH_SHORT).show()
            }
        })

        loginViewModel.userData.observe(this, Observer { userData ->
            if(userData != null){
                userId = userData
                navigateToHomeFragment()
                Log.d("MainActivity", "User data: $userId")
            }
        })


       binding.login.setOnClickListener {
            val email = binding.username.text.toString()
            val password = binding.password.text.toString()
           // val documentId = "Cj4A6RZwrUNBxPYencyH"
            loginViewModel.authenticateUser(email, password)
        }

       binding.register.setOnClickListener{
            llayout.visibility = View.VISIBLE
        }
        binding.submit.setOnClickListener {
            val fName = binding.fName.text.toString()
            val lName = binding.lName.text.toString()
            val email = binding.email.text.toString()
            val pass = binding.newPass.text.toString()
            val user = hashMapOf(
                "first" to fName,
                "last" to lName,
                "email" to email,
                "pass" to pass)

            Log.d("MainActivity", "submit button clicked")
            loginViewModel.registerUser(user)

        }



    }

    private fun navigateToHomeFragment() {
        if(userId !=null && findViewById<FrameLayout>(R.id.fragment_container) != null){
            val homeFragment = HomeFragment.newInstance()
            supportFragmentManager.beginTransaction()
                .replace(R.id.fragment_container, homeFragment)
                .commit()
        }
    }


}
